File: cogs/verify.py
# /verify command — manual license verification via state DOI scrapers
import logging
from datetime import datetime, timezone

import discord
from discord import app_commands
from discord.ext import commands

from config import settings
from db import LicenseDB
from roles import assign_role, remove_role
from scrapers import get_scraper
from sms import send_license_alert

logger = logging.getLogger(__name__)


class VerifyCog(commands.Cog, name="Verify"):
    """/verify slash command for agents to verify their license."""

    def __init__(self, bot: commands.Bot, db: LicenseDB) -> None:
        self.bot = bot
        self.db = db

    @app_commands.command(
        name="verify",
        description="Verify your insurance license and get promoted to Licensed Agent",
    )
    @app_commands.describe(
        first_name="Your legal first name (as on your license)",
        last_name="Your legal last name (as on your license)",
        state="Your home state (2-letter code, e.g. FL, TX, CA)",
        phone="Your phone number for license update texts (optional)",
    )
    async def verify_license(
        self,
        interaction: discord.Interaction,
        first_name: str,
        last_name: str,
        state: str = None,
        phone: str = None,
    ) -> None:
        await interaction.response.defer(ephemeral=True)
        member = interaction.user
        logger.info(f"License verify: {first_name} {last_name} ({state}) by {member}")

        # Pull state from DB if not provided
        if not state:
            agent = await self.db.get_agent(member.id)
            state = agent.get("home_state", "") if agent else ""

        if not state or len(state) != 2:
            await interaction.followup.send(
                "\u274c Please provide your 2-letter state code.\n"
                "Example: `/verify first_name:John last_name:Doe state:FL`",
                ephemeral=True,
            )
            return

        state = state.upper().strip()

        # Save phone number if provided
        if phone:
            clean_phone = "".join(c for c in phone if c.isdigit())
            if len(clean_phone) == 10:
                clean_phone = "+1" + clean_phone
            elif len(clean_phone) == 11 and clean_phone.startswith("1"):
                clean_phone = "+" + clean_phone
            if clean_phone:
                await self.db.update_agent(member.id, phone_number=clean_phone)

        # Run the scraper for this state
        scraper = get_scraper(state)
        try:
            results = await scraper.lookup_by_name(first_name, last_name)
        finally:
            await scraper.close()

        # Find best match — prefer life insurance licenses
        match = None
        for r in results:
            if r.found and r.active:
                if r.is_life_licensed or not r.license_type:
                    match = r
                    break
        if not match:
            for r in results:
                if r.found and r.active:
                    match = r
                    break

        if match and match.found and match.active:
            # ── SUCCESS ──
            await self.db.update_agent(
                member.id,
                npn=match.npn or None,
                license_number=match.license_number or None,
                resident_state=state,
                license_status="licensed",
                license_verified=1,
                license_expiry=match.expiration_date or None,
                verified_at=datetime.now(timezone.utc).isoformat(),
                last_license_check=datetime.now(timezone.utc).isoformat(),
                current_stage=5,
            )
            await self.db.log_activity(
                member.id, "license_verified",
                f"Verified via {state} DOI: {match.full_name} | "
                f"License: {match.license_number} | Status: {match.status}"
            )
            await self.db.log_check(member.id, state, "verified", match.status)

            # Role changes
            try:
                guild = self.bot.get_guild(settings.GUILD_ID)
                guild_member = guild.get_member(member.id) if guild else None
                if guild_member:
                    if settings.LICENSED_AGENT_ROLE_ID:
                        await assign_role(guild_member, settings.LICENSED_AGENT_ROLE_ID, "License verified")
                    if settings.STUDENT_ROLE_ID:
                        await remove_role(guild_member, settings.STUDENT_ROLE_ID, "Promoted")
            except Exception as e:
                logger.warning(f"Role change failed: {e}")

            # Post to channel
            await self._post_verification(member, match, state)

            # SMS confirmation
            agent = await self.db.get_agent(member.id)
            if agent and agent.get("phone_number"):
                await send_license_alert(
                    agent["phone_number"],
                    f"{first_name} {last_name}",
                    state, "verified",
                )

            # DM next steps
            await self._dm_next_steps(member)

            await interaction.followup.send(
                f"\u2705 **License Verified!**\n\n"
                f"**Name:** {match.full_name}\n"
                f"**License #:** {match.license_number or 'N/A'}\n"
                f"**NPN:** {match.npn or 'N/A'}\n"
                f"**State:** {state}\n"
                f"**Status:** {match.status}\n"
                f"**Type:** {match.license_type or 'N/A'}\n\n"
                f"You've been promoted to **Licensed Agent**!",
                ephemeral=True,
            )

        elif results and results[0].error:
            await interaction.followup.send(
                f"\u26a0\ufe0f **Lookup Error:** {results[0].error}\n\n"
                f"The {state} lookup may be temporarily unavailable. Try again later.",
                ephemeral=True,
            )
        else:
            await interaction.followup.send(
                "\u274c **Could not verify your license.**\n\n"
                "\u2022 Your name may not match what's on file\n"
                "\u2022 Your license may not be processed yet\n"
                "\u2022 You may be licensed in a different state\n\n"
                f"**Searched:** {first_name} {last_name} in {state}\n\n"
                "Contact your upline for manual verification.",
                ephemeral=True,
            )

    # ── /license-history — check your own verification history ──

    @app_commands.command(
        name="license-history",
        description="View your license check history",
    )
    async def license_history(self, interaction: discord.Interaction) -> None:
        await interaction.response.defer(ephemeral=True)

        checks = await self.db.get_check_history(interaction.user.id, limit=5)
        if not checks:
            await interaction.followup.send(
                "No license checks on file yet. Run `/verify` to get started!",
                ephemeral=True,
            )
            return

        lines = []
        for c in checks:
            emoji = "\u2705" if c["status"] == "verified" or c["status"] == "active" else "\u274c"
            lines.append(
                f"{emoji} **{c['state']}** — {c['status'].title()} "
                f"({c['check_date'][:10]})"
            )

        embed = discord.Embed(
            title="\U0001f4cb  License Check History",
            description="\n".join(lines),
            color=0x3498DB,
        )
        await interaction.followup.send(embed=embed, ephemeral=True)

    # ── Helpers ──

    async def _post_verification(self, user, match, state: str) -> None:
        channel_id = settings.LICENSE_CHECK_CHANNEL_ID or settings.HIRING_LOG_CHANNEL_ID
        channel = self.bot.get_channel(channel_id) if channel_id else None
        if not channel:
            return
        embed = discord.Embed(
            title="\U0001f393  License Verified!",
            description=(
                f"<@{user.id}> verified as a licensed agent.\n\n"
                f"**Name:** {match.full_name}\n"
                f"**License #:** {match.license_number or 'N/A'}\n"
                f"**State:** {state} | **Status:** {match.status}"
            ),
            color=0x2ECC71,
            timestamp=datetime.now(timezone.utc),
        )
        await channel.send(embed=embed)

    async def _dm_next_steps(self, member) -> None:
        try:
            embed = discord.Embed(
                title="\U0001f389  License Verified! Next Step: Contracting",
                description=(
                    f"Welcome **{member.display_name}**! Your license has been "
                    f"confirmed. Use `/contract` in the server to book your "
                    f"contracting appointment."
                ),
                color=0x9B59B6,
            )
            embed.add_field(
                name="\U0001f4dd  What to Prepare",
                value=(
                    "\u2022 Government-issued photo ID\n"
                    "\u2022 Social Security number\n"
                    "\u2022 E&O insurance info\n"
                    "\u2022 Bank info for direct deposit\n"
                    "\u2022 Resident state license number"
                ),
                inline=False,
            )
            await member.send(embed=embed)
        except discord.Forbidden:
            logger.warning(f"Cannot DM {member}")


async def setup(bot: commands.Bot) -> None:
    await bot.add_cog(VerifyCog(bot, bot._db))

File: bot/verify.go
package bot

import (
	"context"
	"fmt"
	"log"
	"strings"
	"time"

	"github.com/bwmarrin/discordgo"

	"license-bot-go/api/websocket"
	"license-bot-go/db"
	"license-bot-go/scrapers"
)

func (b *Bot) handleVerify(s *discordgo.Session, i *discordgo.InteractionCreate) {
	if i.Member == nil {
		s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
			Type: discordgo.InteractionResponseChannelMessageWithSource,
			Data: &discordgo.InteractionResponseData{
				Content: "This command can only be used in a server.",
				Flags:   discordgo.MessageFlagsEphemeral,
			},
		})
		return
	}

	// Step 1: Defer (ephemeral)
	err := s.InteractionRespond(i.Interaction, &discordgo.InteractionResponse{
		Type: discordgo.InteractionResponseDeferredChannelMessageWithSource,
		Data: &discordgo.InteractionResponseData{
			Flags: discordgo.MessageFlagsEphemeral,
		},
	})
	if err != nil {
		log.Printf("Defer failed: %v", err)
		return
	}

	// Step 2: Extract options
	opts := i.ApplicationCommandData().Options
	optMap := make(map[string]string)
	for _, opt := range opts {
		optMap[opt.Name] = opt.StringValue()
	}

	firstName := optMap["first_name"]
	lastName := optMap["last_name"]
	state := strings.ToUpper(strings.TrimSpace(optMap["state"]))
	phone := optMap["phone"]

	userID := i.Member.User.ID
	userIDInt, err := parseDiscordID(userID)
	if err != nil {
		b.followUp(s, i, "Internal error. Please try again.")
		return
	}
	guildIDInt, err := parseDiscordID(i.GuildID)
	if err != nil {
		b.followUp(s, i, "Internal error. Please try again.")
		return
	}

	log.Printf("License verify: %s %s (%s) by %s", firstName, lastName, state, userID)

	// Step 3: Pull state from DB if not provided
	if state == "" {
		agent, err := b.db.GetAgent(context.Background(), userIDInt)
		if err == nil && agent != nil {
			state = agent.State
		}
	}

	if state == "" || len(state) != 2 {
		b.followUp(s, i, "Please provide your 2-letter state code.\nExample: `/verify first_name:John last_name:Doe state:FL`")
		return
	}

	// Step 4: Save phone number if provided
	if phone != "" {
		cleanPhone := cleanPhoneNumber(phone)
		if cleanPhone != "" {
			b.db.UpsertAgent(context.Background(), userIDInt, guildIDInt, db.AgentUpdate{
				PhoneNumber: &cleanPhone,
			})
		}
	}

	// Step 5: Get scraper and lookup
	scraper := b.registry.GetScraper(state)

	ctx, cancel := context.WithTimeout(context.Background(), 90*time.Second)
	defer cancel()

	results, err := scraper.LookupByName(ctx, firstName, lastName)
	if err != nil {
		log.Printf("Scraper error for %s: %v", state, err)
		msg := fmt.Sprintf("The %s license lookup is temporarily unavailable. Please try again later.", state)
		if url := scraper.ManualLookupURL(); url != "" {
			msg += fmt.Sprintf("\n\nManual lookup: %s", url)
		}
		b.followUp(s, i, msg)
		return
	}

	// Step 6: Find best match
	var match *scrapers.LicenseResult

	// First pass: prefer life-licensed active results
	for idx := range results {
		if results[idx].Found && results[idx].Active && results[idx].IsLifeLicensed() {
			match = &results[idx]
			break
		}
	}
	// Second pass: any active result
	if match == nil {
		for idx := range results {
			if results[idx].Found && results[idx].Active {
				match = &results[idx]
				break
			}
		}
	}

	// Step 7: Respond based on result
	if match != nil {
		// SUCCESS
		verified := true
		stage := db.StageVerified
		b.db.UpsertAgent(context.Background(), userIDInt, guildIDInt, db.AgentUpdate{
			FirstName:       &firstName,
			LastName:        &lastName,
			State:           &state,
			LicenseVerified: &verified,
			LicenseNPN:      &match.NPN,
			CurrentStage:    &stage,
		})

		b.db.SaveLicenseCheck(context.Background(), db.LicenseCheck{
			DiscordID:      userIDInt,
			GuildID:        guildIDInt,
			FirstName:      firstName,
			LastName:        lastName,
			State:          state,
			NPN:            match.NPN,
			LicenseNumber:  match.LicenseNumber,
			LicenseType:    match.LicenseType,
			LicenseStatus:  match.Status,
			ExpirationDate: match.ExpirationDate,
			LOAs:           match.LOAs,
			Found:          true,
		})

		// Broadcast license_verified event
		b.publishEvent(websocket.EventLicenseVerified, websocket.LicenseVerifiedData{
			DiscordID: userID,
			State:     state,
			FullName:  firstName + " " + lastName,
			NPN:       match.NPN,
		})
		b.publishEvent(websocket.EventStageChanged, websocket.StageChangedData{
			DiscordID: userID,
			NewStage:  db.StageVerified,
			ChangedBy: "verify_command",
		})

		// Fire-and-forget secondary actions
		go b.postToChannel(s, i, match, state)
		go b.dmNextSteps(s, i, match, state)
		go b.assignRoles(s, i)
		go b.syncGHLStage(userIDInt, db.StageVerified)

		b.followUp(s, i, fmt.Sprintf(
			"**License Verified!**\n\n"+
				"**Name:** %s\n"+
				"**License #:** %s\n"+
				"**NPN:** %s\n"+
				"**State:** %s\n"+
				"**Status:** %s\n"+
				"**Type:** %s\n\n"+
				"You've been promoted to **Licensed Agent**!",
			nvl(match.FullName, "N/A"),
			nvl(match.LicenseNumber, "N/A"),
			nvl(match.NPN, "N/A"),
			state,
			nvl(match.Status, "N/A"),
			nvl(match.LicenseType, "N/A"),
		))

	} else if len(results) > 0 && results[0].Error != "" {
		// Error from scraper (e.g., manual URL fallback)
		b.followUp(s, i, fmt.Sprintf("**Lookup Error:** %s\n\nThe %s lookup may be temporarily unavailable. Try again later.", results[0].Error, state))

	} else {
		// Not found
		msg := fmt.Sprintf(
			"**Could not verify your license.**\n\n"+
				"- Your name may not match what's on file\n"+
				"- Your license may not be processed yet\n"+
				"- You may be licensed in a different state\n\n"+
				"**Searched:** %s %s in %s\n\n"+
				"Contact your upline for manual verification.",
			firstName, lastName, state,
		)
		if url := scraper.ManualLookupURL(); url != "" {
			msg += fmt.Sprintf("\n\nManual lookup: %s", url)
		}
		b.followUp(s, i, msg)
	}
}

func (b *Bot) followUp(s *discordgo.Session, i *discordgo.InteractionCreate, content string) {
	_, err := s.FollowupMessageCreate(i.Interaction, true, &discordgo.WebhookParams{
		Content: content,
		Flags:   discordgo.MessageFlagsEphemeral,
	})
	if err != nil {
		log.Printf("Follow-up failed: %v", err)
	}
}

// verifyLogChannelID returns the channel ID for posting verification results.
// Falls back: LicenseVerifyLogChannelID -> LicenseCheckChannelID -> HiringLogChannelID.
func (b *Bot) verifyLogChannelID() string {
	if b.cfg.LicenseVerifyLogChannelID != "" {
		return b.cfg.LicenseVerifyLogChannelID
	}
	if b.cfg.LicenseCheckChannelID != "" {
		return b.cfg.LicenseCheckChannelID
	}
	return b.cfg.HiringLogChannelID
}

func (b *Bot) postToChannel(s *discordgo.Session, i *discordgo.InteractionCreate, match *scrapers.LicenseResult, state string) {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("postToChannel panic: %v", r)
		}
	}()

	channelID := b.verifyLogChannelID()
	if channelID == "" {
		return
	}

	userID := i.Member.User.ID
	embed := &discordgo.MessageEmbed{
		Title: "License Verified!",
		Description: fmt.Sprintf(
			"<@%s> verified as a licensed agent.\n\n"+
				"**Name:** %s\n"+
				"**License #:** %s\n"+
				"**State:** %s | **Status:** %s",
			userID, match.FullName, nvl(match.LicenseNumber, "N/A"), state, match.Status,
		),
		Color:     0x2ECC71,
		Timestamp: time.Now().Format(time.RFC3339),
	}

	s.ChannelMessageSendEmbed(channelID, embed)
}

func (b *Bot) dmNextSteps(s *discordgo.Session, i *discordgo.InteractionCreate, match *scrapers.LicenseResult, state string) {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("dmNextSteps panic: %v", r)
		}
	}()

	channel, err := s.UserChannelCreate(i.Member.User.ID)
	if err != nil {
		log.Printf("Cannot create DM channel: %v", err)
		return
	}

	// Build license info fields
	var fields []*discordgo.MessageEmbedField

	fields = append(fields, &discordgo.MessageEmbedField{
		Name:   "Full Name",
		Value:  nvl(match.FullName, "N/A"),
		Inline: true,
	})
	fields = append(fields, &discordgo.MessageEmbedField{
		Name:   "State",
		Value:  state,
		Inline: true,
	})
	fields = append(fields, &discordgo.MessageEmbedField{
		Name:   "Status",
		Value:  nvl(match.Status, "N/A"),
		Inline: true,
	})
	fields = append(fields, &discordgo.MessageEmbedField{
		Name:   "License #",
		Value:  nvl(match.LicenseNumber, "N/A"),
		Inline: true,
	})
	fields = append(fields, &discordgo.MessageEmbedField{
		Name:   "NPN",
		Value:  nvl(match.NPN, "N/A"),
		Inline: true,
	})
	fields = append(fields, &discordgo.MessageEmbedField{
		Name:   "License Type",
		Value:  nvl(match.LicenseType, "N/A"),
		Inline: true,
	})

	if match.IssueDate != "" {
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   "Issue Date",
			Value:  match.IssueDate,
			Inline: true,
		})
	}
	if match.ExpirationDate != "" {
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   "Expiration Date",
			Value:  match.ExpirationDate,
			Inline: true,
		})
	}
	if match.Resident {
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   "Residency",
			Value:  "Resident",
			Inline: true,
		})
	}
	if match.LOAs != "" {
		// Truncate LOAs if too long for embed field (max 1024 chars)
		loas := match.LOAs
		if len(loas) > 900 {
			loas = loas[:900] + "..."
		}
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   "Lines of Authority",
			Value:  loas,
			Inline: false,
		})
	}
	if match.BusinessAddress != "" {
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   "Business Address",
			Value:  match.BusinessAddress,
			Inline: false,
		})
	}
	if match.BusinessPhone != "" {
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   "Business Phone",
			Value:  match.BusinessPhone,
			Inline: true,
		})
	}
	if match.Email != "" {
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   "Email",
			Value:  match.Email,
			Inline: true,
		})
	}
	if match.County != "" {
		fields = append(fields, &discordgo.MessageEmbedField{
			Name:   "County",
			Value:  match.County,
			Inline: true,
		})
	}

	// Add next steps section
	fields = append(fields, &discordgo.MessageEmbedField{
		Name: "\u200b\nNext Step: Contracting",
		Value: "Use `/contract` in the server to book your contracting appointment.\n\n" +
			"**What to Prepare:**\n" +
			"- Government-issued photo ID\n" +
			"- Social Security number\n" +
			"- E&O insurance info\n" +
			"- Bank info for direct deposit\n" +
			"- Resident state license number",
		Inline: false,
	})

	embed := &discordgo.MessageEmbed{
		Title:       "License Verified!",
		Description: fmt.Sprintf("Welcome **%s**! Your license has been confirmed. Here are your full license details:", i.Member.User.Username),
		Color:       0x2ECC71,
		Fields:      fields,
		Timestamp:   time.Now().Format(time.RFC3339),
		Footer: &discordgo.MessageEmbedFooter{
			Text: "VIPA License Verification",
		},
	}

	s.ChannelMessageSendEmbed(channel.ID, embed)
}

func (b *Bot) assignRoles(s *discordgo.Session, i *discordgo.InteractionCreate) {
	defer func() {
		if r := recover(); r != nil {
			log.Printf("assignRoles panic: %v", r)
		}
	}()

	if b.cfg.LicensedAgentRoleID != "" {
		err := s.GuildMemberRoleAdd(i.GuildID, i.Member.User.ID, b.cfg.LicensedAgentRoleID)
		if err != nil {
			log.Printf("Failed to add Licensed Agent role: %v", err)
		}
	}

	if b.cfg.StudentRoleID != "" {
		err := s.GuildMemberRoleRemove(i.GuildID, i.Member.User.ID, b.cfg.StudentRoleID)
		if err != nil {
			log.Printf("Failed to remove Student role: %v", err)
		}
	}
}

// VerifyResult is the outcome of a license verification attempt.
type VerifyResult struct {
	Found   bool
	Match   *scrapers.LicenseResult
	Error   string
	Message string
}

// performVerification runs a license lookup and returns the result without any Discord interaction.
func (b *Bot) performVerification(ctx context.Context, firstName, lastName, state string, discordID, guildID int64) VerifyResult {
	if state == "" || len(state) != 2 {
		return VerifyResult{Error: "invalid state code"}
	}

	scraper := b.registry.GetScraper(state)

	results, err := scraper.LookupByName(ctx, firstName, lastName)
	if err != nil {
		msg := fmt.Sprintf("Lookup error for %s: %v", state, err)
		return VerifyResult{Error: msg}
	}

	// Find best match: prefer life-licensed active results
	var match *scrapers.LicenseResult
	for idx := range results {
		if results[idx].Found && results[idx].Active && results[idx].IsLifeLicensed() {
			match = &results[idx]
			break
		}
	}
	if match == nil {
		for idx := range results {
			if results[idx].Found && results[idx].Active {
				match = &results[idx]
				break
			}
		}
	}

	if match != nil {
		// Save to DB
		verified := true
		stage := db.StageVerified
		b.db.UpsertAgent(ctx, discordID, guildID, db.AgentUpdate{
			FirstName:       &firstName,
			LastName:        &lastName,
			State:           &state,
			LicenseVerified: &verified,
			LicenseNPN:      &match.NPN,
			CurrentStage:    &stage,
		})
		b.db.SaveLicenseCheck(ctx, db.LicenseCheck{
			DiscordID:      discordID,
			GuildID:        guildID,
			FirstName:      firstName,
			LastName:       lastName,
			State:          state,
			NPN:            match.NPN,
			LicenseNumber:  match.LicenseNumber,
			LicenseType:    match.LicenseType,
			LicenseStatus:  match.Status,
			ExpirationDate: match.ExpirationDate,
			LOAs:           match.LOAs,
			Found:          true,
		})
		return VerifyResult{Found: true, Match: match}
	}

	if len(results) > 0 && results[0].Error != "" {
		return VerifyResult{Error: results[0].Error}
	}

	return VerifyResult{
		Found:   false,
		Message: fmt.Sprintf("No active license found for %s %s in %s", firstName, lastName, state),
	}
}

func cleanPhoneNumber(phone string) string {
	digits := strings.Map(func(r rune) rune {
		if r >= '0' && r <= '9' {
			return r
		}
		return -1
	}, phone)
	// Remove leading 1 for US numbers
	if len(digits) == 11 && digits[0] == '1' {
		digits = digits[1:]
	}
	// Require exactly 10 digits for a valid US phone number
	if len(digits) != 10 {
		return ""
	}
	return "+1" + digits
}

func nvl(s, fallback string) string {
	if s == "" {
		return fallback
	}
	return s
}
